### PR TITLE
[AMDGPU][test]added unique and sort options for update_mc_test_check script

### PR DESCRIPTION
**Pull request by broxigarchen**
add a unique and a sort option to the update_mc_test_check script.

These mc asm/dasm files are usually large in number of lines, and these lines are mostly similar to each other. These options can be useful when maintainer is merging or resolving conflicts by making the file identifical

Also fixed a small issue in asm/dasm such that the auto generated header line is 
1. asm using ";" instead of "//" as comment marker
2. dasm using ";" instead of "#" as comment marker

### CODE REVIEW

**Review by arichardson — COMMENTED** [review]
Could you add a test for sorting with an input that contains some comments. 

Ideally those comments would remaing attached to the line they were on before.

**Comment by broxigarchen** [comment]
> Could you add a test for sorting with an input that contains some comments.
> 
> Ideally those comments would remaing attached to the line they were on before.

added a `sort_with_comment ` test

The comment within the sameline or within the next continued line will work
`
TESTLINE  // comment

TESTLINE
// coment
// coment
`

For these cases it will not
`
 // comment
TESTLINE

TESTLINE

// coment
`

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:348]: Shouldn't this try to skip leading comments? For cases like:

```
// FIXME: This test is incorrect because r128 assumes a 128-bit RSRC.
image_load v0, v0, s[0:7] dmask:0x1 dim:SQ_RSRC_IMG_1D r128
// GFX12: encoding: [0x10,0x00,0x40,0xd0,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
```

**Review by kosarev — COMMENTED** [review]
What is your idea about sorting all-comments cases, like:

```
// FIXME: For atomic instructions, glc must be placed immediately following
// the data regiser.  These forms aren't currently supported:
// FIXME: offset:0 required
// flat_atomic_add v1, v[3:4], v5 slc glc
```

?

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:327]: So does this mean we actually only allow unique lines in the output with this option? How will this work with multiple cases having identical comments?

I think this option should aim to remove identical test cases rather than individual lines? Probably with duplicates reported regardless of whether this option is used?

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:133]: What are these unexpected lines?

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:131]: I guess not having this enabled by default might be problematic. Imagine some test cases added without sorting at random positions in one patch and then sorting them in another patch as some more test cases are added. This would also move cases added in the first patch. (I assume the idea is that ultimately we want keep all MC test files sorted at all times.)

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:327]: It only remove duplicated test line. If there are same comments they should stay.

I can add a case for this in the test

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:348]: I haven't add this in the first place but yeah I think this can be added

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:133]: The sorting is not very smart to handle all cases, and there are cases that might suprise the user, for example, the --sort and --remove-duplicate used at same time:

```
testlineA
//comment 1

testlineA
//commet 2
```
after remove-duplicate it become
```
testlineA
//comment 1

//commet 2
```

and then sort, `comment2` will be moved to other places which seems coming out of no where.

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:131]: The same reason as above. I think it's better asking user to run --sort manually.

I think we can add a hint by checking the auto-generated header line
`# auto updated by update_mc_test_check.py UTC_ARGS: --sort`

if there is a `--sort` there but there is no `--sort` being passed by the arguments, we can shot a warning

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [reply llvm/utils/update_mc_test_checks.py:133]: This smells more like a bug rather than a limitation. We may report cases that only differ by their comments, but surely we shouldn't remove input lines from test cases and leave just comments?

Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:133]: I think this might be hard to determine. i.e. the `//comment 2` can be a comment for the whole section bellow, and it should be kept even though the first testline is removed as a duplication. Thus I tried removing minimum lines when removing duplications and leave this to the user.

What do you think?

File: llvm/utils/update_mc_test_checks.py
Comment (jayfoad) [inline llvm/utils/update_mc_test_checks.py:123]: Maybe just call it `--unique` (by vague analogy with Unix commands `uniq` and `sort -u`).

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [reply llvm/utils/update_mc_test_checks.py:133]: > the `//comment 2` can be a comment for the whole section bellow,

If the script reported these duplicate cases (established by considering them with all comment lines stripped), then I think it shouldn't be a problem. The user will see the warning/error, fix the test case and run the script again.

I'd expect the script to play safe and only remove test cases when there's no doubt that they are duplicates, that is, when they are absolutely identical (which I guess should be the case for most of duplicates).

Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:135]: updated

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py]: added. Now comment line are skipped

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py]: good idea. Use unique now

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py]: added.

duplicated comment should not be dropped

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected
Comment (Sisyph) [inline llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected:5]: I can't tell from this test case. Does the disassembler sort on the asm output, or the hex input? I would think the asm output would be preferable

Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected:5]: For now all sort base on the input text so for dasm it sorts on the hex input

If we want to sort base on the asm output, what should we do with the error case?

**Comment by broxigarchen** [comment]
Adding more reviewers to seek a wider consensus on sorting MC tests throughout

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected
Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected:5]: Hey Joe I checked the dasm and it seems sorting with hex mostly cannot group same instruction to same place, since the first byte are gprs

If we want to sort on output, it probably a bit over complicated seems we may have multiple out asm mixed with errors.

What about passing a sorting mask? It seems most of the dasm file has same length of bytes. Assume we pass a sorting mask with "0xFF0000", then for each line, i.e. "0xa,0xb,0xc", the script can merge them to 0x0a0b0c then use the mask to get the key for sorting. Do you think this could be a ugly solution? also @kosarev

Comment (Sisyph) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected:5]: Perhaps you could exclude disasm file sorting from this patch? It seems like the other cases are working well, and we don't need to do it all at once. I don't see how sorting on the hex is going to be useful. The bits that encode the opcode would make sense to sort on, but that varies so much from instruction to instruction it will be horrible to maintain. With sorting on the output, can you redirect errors to null? Or maybe we separate those test files with both passing and error tests?

**Comment by broxigarchen** [comment]
> Adding more reviewers to seek a wider consensus on sorting MC tests throughout

Discussed offline with Joe and Ivan. The dasm sorting might not be very useful right now when sorted with hex input, and we might not need it after dasm test are merged into asm test.

For this patch, drop the support of sorting for dasm test. Added a error message so that script error out if dasm test is passed with --sort

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected
Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_dasm_sort.txt.expected]: done. dasm sorting is removed from this patch

**Comment by broxigarchen** [comment]
ping!

**Review by Sisyph — APPROVED** [review]
LGTM, but please wait for @kosarev

**Comment by Sisyph** [comment]
Since this script is now being used for in tree tests (https://github.com/llvm/llvm-project/commit/b27acebe63b128feb3bb9c3c62d77f235d2e6a6e) , I would recommend landing it now. If anything arises in post commit review, please re-run the script on the affected tests.

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:370]: Nit: either the `else` or the `continue` not needed?

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:354]: Nit: can be just `if`.

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:404]: Can this be simplified to something like:

```suggestion
                def get_key(line):
                    # get key string to be compared
                    if line in test_dic:
                       line = test_dic[line]
                    else:
                        line = line.split("\n")[0]
                    return (not isRunLine(line), line)

                test_units = sorted(test_units, key=get_key)
```

?

**Review by kosarev — APPROVED** [review]
LGTM with nits.

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py]: Yup I think this can be simplified.

I thought I have to pass a comparator here but seems just a get key could work. Nice suggestion!

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py]: done

File: llvm/utils/update_mc_test_checks.py
Comment (kosarev) [inline llvm/utils/update_mc_test_checks.py:9]: This is unused now.

Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:9]: I'll remove this in another patch

**Pull request merged**
